src/Board.cpp

```diff
@@ -0,0 +1,207 @@
+ *                      41 - the last movement was legal and cause check
+ *                      42 - the last movement was legal
+ *
+ * @param move the move to check
+ * @return int the response code
+ */
+int Board::getResponeCode(const std::string &move) {
+    Position src = {move[0] - 'a', move[1] - '1'};
+    Position dst = {move[2] - 'a', move[3] - '1'};
+
+    if(isEmpty(src)) return 11;
+    if(isEnemy(src)) return 12;
+    if(isCurPlayerPiece(dst)) return 13;
+    if(!m_board[src.row][src.col]->isValidMove(src, dst) ||
+        isPathBlocked(src, dst)) return 21;
+    if(isMoveCauseYouCheck(src, dst)) return 31;
+    updateBoard(src, dst);
+    if(isCheckmate()) return 41;
+    return 42;
+}
+
+/**
+ * @brief check if the cell is empty
+ *
+ * @param src the source position
+ * @return true if the cell is empty
+ * @return false if the cell is not empty
+ */
+bool Board::isEmpty(const Position src) const {
+    return m_board[src.row][src.col] == nullptr;
+}
+
+/**
+ * @brief check if the cell is enemy piece
+ *
+ * @param src the source position
+ * @return true if the cell is enemy piece
+ * @return false if the cell is not enemy piece
+ */
+bool Board::isEnemy(const Position src) const {
+    return m_board[src.row][src.col]->isWhite() != m_isWhiteTurn;
+}
+
+/**
+ * @brief check if the cell is current player piece
+ *
+ * @param src the source position
+ * @return true if the cell is current player piece
+ * @return false if the cell is not current player piece
+ */
+bool Board::isCurPlayerPiece(const Position src) const {
+    return !isEmpty(src) && m_board[src.row][src.col]->isWhite() == m_isWhiteTurn;
+}
+
+/**
+ * @brief check if the path between the source and the destination is blocked
+ * @param source
+ * @param dest
+ * @return true if the path is blocked
+ * @return false if the path is not blocked
+ */
+bool Board::isPathBlocked(const Position& source, const Position& dest) const
+{
+    int dx = dest.row - source.row;
+    int dy = dest.col - source.col;
+
+    // Movement is not straight or diagonal
+    if (dx != 0 && dy != 0 && dx != dy)
+        return false;
+
+    // Check if there is a piece between the source and the destination
+    int steps = std::max(std::abs(dx), std::abs(dy));
+    int x_step = (dx > 0) ? 1 : ((dx < 0) ? -1 : 0);
+    int y_step = (dy > 0) ? 1 : ((dy < 0) ? -1 : 0);
+    Position curPos = source;
+    for (int i = 1; i < steps; i++) {
+        curPos.row += x_step;
+        curPos.col += y_step;
+        if (!isEmpty(curPos))
+            return true;
+    }
+    return false;
+}
+
+/**
+ * @brief update the board after the move
+ * @param src the source position
+ * @param des the destination position
+ */
+void Board::updateBoard(const Position src, const Position des) {
+    if(isKing(src))
+        m_kingPos[m_board[src.row][src.col]->isWhite()] = des;
+    if(isPawn(src))
+        dynamic_cast<Pawn*>(m_board[src.row][src.col].get())->setFirstMove();
```

Comment:
Since you already know that m_board[src.row][src.col] is a Pawn*, you can use static_cast instead of dynamic_cast. This will improve performance as static_cast performs a compile-time type check instead of a runtime check.